vncviewer/Viewport.cxx

```diff
@@ -598,32 +600,52 @@ int Viewport::handle(int event)
   case FL_DRAG:
   case FL_MOVE:
   case FL_MOUSEWHEEL:
-    buttonMask = 0;
+
+    // Reset state for left, right, middle
+    mouseButtonMask &= ~(0x7);
```

Comment:
Unfortunately, FLTK is quite buggy and I couldn't reliably keep the button states in sync:
  * https://github.com/fltk/fltk/issues/1076
  * https://github.com/fltk/fltk/issues/1077
  
So we either have to fix these issues upstream, or work around this issue by looking at the button states from LMB/MMB/RMB for each event (current implementation).

The bug that is in FLTK means that we can get in a scenario where the state for the forward/back buttons get out of sync in some edge cases. I could reproduce this by holding the back button, then tabbing to xterm and releasing the back button. Hovering back to the viewport, the `mouseButtonMask` had the back button bit set. I had to click the back button once to reset it.

Reply:
@CendioHalim wrote:
> Unfortunately, FLTK is quite buggy and I couldn't reliably keep the button states in sync ...

I disagree with **this wording** (_"quite buggy"_).

I commented on both above mentioned FLTK issues. Fact is that FLTK **does not support** "mouse buttons 8 and 9" (yet). These buttons are AFAICT not at all documented in the Xorg docs and in any X11 headers! I'm also not aware of any other docs (conventions) that define the meaning and internals of these extended mouse buttons.

Dragging the mouse while holding an unsupported mouse button **and** using `Alt+Tab` - which is usually intercepted/handled by the OS (WM) - can't be considered a bug as well. As you, @CendioHalim, wrote elsewhere, X11 does not keep the button states itself, and FLTK relies on X11 sending the current button states.

Of course we're interested in extending FLTK to support these additional mouse buttons, but calling the current state "quite buggy" is not appropriate.

Reply:
> @CendioHalim wrote:
> 
> > Unfortunately, FLTK is quite buggy and I couldn't reliably keep the button states in sync ...
> 
> I disagree with **this wording** (_"quite buggy"_).

I apologize for my wording.  

FLTK sends input events for back/forward, and I've found some cases where this happens incorrectly. Whether this is a bug in FLTK or the WM/OS, I'm not sure and here is not much I can do from my perspective.

FWIW, at one point, I added my own system handler to handle all mouse events with native code and not rely on FLTK. With that setup, I did not have the issue described in my earlier comment. So there is some kind of weird interaction going on in that particular case.

> Dragging the mouse while holding an unsupported mouse button and using Alt+Tab - which is usually intercepted/handled by the OS (WM) - can't be considered a bug as well. 

To clarify, this happens when using LMB as well. The bits I'm talking about in my comment is my own button state tracker, not anything from FLTK. Both `Fl::event_state()` and `fl_xevent->xbutton.state` are set correctly, but I still get `FL_PUSH` when no buttons are pressed.

Reply:
@CendioHalim wrote:
> I apologize for my wording.

Thanks, appreciated.

Regarding status of back/forward mouse buttons when switching context (`Alt+Tab` or `unfocus/focus`): ISTR that FLTK sets up the current keyboard and mouse button status whenever we [lose and] get focus [back]. On X11 I'm pretty sure that we read the current status from the X11 (`FOCUS` or any other) event which would preclude getting the back/forward button status since X11 doesn't keep the status and thus doesn't send it in any event.

OTOH, if we kept the status internally and the user pushed and held one of these buttons before context switch (i.e. focus loss) we wouldn't be able to determine the status when we get the focus back because the user could have released the button after the switch. I think this is one scenario you mentioned previously here or elsewhere.

This is something that needs investigation and can maybe not be solved at all unless someone finds out how to query the current button status when we get the focus back.

PS: I received my new mouse and did a first few tests. Back/forward buttons work in firefox as expected. I'll post my results later at one of the FLTK issues or PR's.